### PR TITLE
Fix for #1585. Improved EntityDescriptor caching.

### LINKED ISSUE #446: Fluent ....HasTableName(..) works only first time (version 1.0.7.4)
Seems like changing mapping TableName with fluent mapping works only once per process instance.

Example:
I have a set of tables with same data models. 

```
        // Data model class (for different tables):
        public class SomeModel
        {
            [Column("id")]
            public int Id { get; set; }
            [Column("field")]
            public string Field { get; set; }
        }

        // Testing mapping same class with different tables (in different instances of DbConnection!)
        public void Test()
        {
            var confName = "Test";

            // map to table1
            using (var db = new LinqToDB.Data.DataConnection(confName))
            {
                db.MappingSchema.GetFluentMappingBuilder().Entity<ModelBase>().HasTableName("table1");
                var q = db.GetTable<ModelBase>(); // = sql=... from "table1"...
                var t = q.ToArray();  // <--- mapped to "table1" 
            }
            // map to table2
            using (var db = new LinqToDB.Data.DataConnection(confName))
            {
                db.MappingSchema.GetFluentMappingBuilder().Entity<ModelBase>().HasTableName("table2");
                var q = db.GetTable<ModelBase>(); // = sql=... from "table1"... !!
                var t = q.ToArray();  // <--- still mapped to "table1", not "table2"  !!!
            }
        }

```

I am using linq2db version 1.0.7.4
Maybe I've missed something?

### LINKED ISSUE #1585: Performance Issue with current Version (GetEntityDescriptor)
We use linq2db with our own MappingSchema wich we use when we create a DataConnection as a Paremeter.

It seems that the GetEntityDescriptor call always creates a new EntityDescriptor, wich takes most of our time. Descriptors should be cached in MappingSchemaInfo, but maybe that is also always recreated?

See Our perf. Profile:

image
image

### PR DESCRIPTION
**Pull request by jogibear9988**
Fixes #1585
Fixes #446

### CODE REVIEW

**Comment by sdanyliv** [comment]
I have prepared fix but it needs merging https://github.com/linq2db/linq2db/pull/1491 first.
For comparing you can review this commit: https://github.com/linq2db/linq2db/pull/1587/commits/9c40dd5bf501c27a260bbb93f9c68561ecbe2c0d

**Comment by jogibear9988** [comment]
For me the fix looks good (if it works ;-) ), but I don't understand why we don't use a Dictionary

**Comment by sdanyliv** [comment]
Because after several days of server work, this dictionary will grab a lot of memory. 
We need some mechanism to purge not needed data. In this case if `EntityDescriptor` is not used 1 hour, it will be purged from cache.

**Comment by jogibear9988** [comment]
But EntityDescriptors normaly will not increase, or am I wrong? Only the DataTypes for wich I created a Mapping. Ore are there other Types wich are stored in the Cache?

**Comment by sdanyliv** [comment]
We are trying to cover many cases. There are users who change MappingSchema dynamically very often and we need to be sure that not needed descriptors are purged.

**Comment by jogibear9988** [comment]
thanks for the info and fix

File: Source/LinqToDB/Metadata/FluentMetadataReader.cs [hunk 1]
```diff
@@ -87,5 +87,10 @@ public void AddAttribute(MemberInfo memberInfo, Attribute attribute)
 		/// <inheritdoc cref="IMetadataReader.GetDynamicColumns"/>
 		public MemberInfo[] GetDynamicColumns(Type type)
 			=> _dynamicColumns.TryGetValue(type, out var dynamicColumns) ? dynamicColumns.Keys.ToArray() : new MemberInfo[0];
+
+		public Type[] GetRegisteredTypes()
```
Comment (MaceWindu) [inline Source/LinqToDB/Metadata/FluentMetadataReader.cs]: xml-doc missing

File: Source/LinqToDB/Mapping/MappingSchema.cs [hunk 2]
```diff
@@ -1461,32 +1459,64 @@ public StringComparer ColumnNameComparer
...
+		/// <returns>
+		/// Mapping types.
+		/// </returns>
+		public Type[] GetDefinedTypes()
+		{
+			return Schemas.SelectMany(s => s.GetRegisteredTypes()).ToArray();
+		}
+
+		/// <summary>
+		/// Clears EntityDescriptor Cache.
```
Comment (MaceWindu) [inline Source/LinqToDB/Mapping/MappingSchema.cs]: `Cache` -> `cache`

File: Source/LinqToDB/Mapping/MappingSchema.cs [hunk 3]
```diff
@@ -1461,32 +1459,64 @@ public StringComparer ColumnNameComparer
...
 		/// <param name="type">Mapped type.</param>
 		/// <returns>Mapping descriptor.</returns>
 		public EntityDescriptor GetEntityDescriptor(Type type)
 		{
-			return Schemas[0].GetEntityDescriptor(this, type);
+			var key = new { Type = type, ConfigurationID };
+			var ed = EntityDescriptorsCache.GetOrCreate(key,
+				o =>
+				{
+					o.SlidingExpiration = Configuration.Linq.CacheSlidingExpiration;
```
Comment (jogibear9988) [inline Source/LinqToDB/Mapping/MappingSchema.cs]: Shouldn't we create a extra Timeout For the entity descriptors?

Comment (sdanyliv) [reply Source/LinqToDB/Mapping/MappingSchema.cs]: Maybe, you can introduce it if you feel uncomfortable with current settings. My opinion it should be enough.

**Comment by jogibear9988** [comment]
is there anything wrong, why the tests fail?

**Comment by jogibear9988** [comment]
performance is much better now !

**Comment by MaceWindu** [comment]
Superseeds #1491? Any objections for merge?

**Comment by jogibear9988** [comment]
@sdanyliv could be merged?

**Pull request merged**